File: sdk/python/src/inferlet/session.py
"""
Session — ``pie:core/session``.

Client communication: send/receive text and files.
"""

from __future__ import annotations

import json as _json
from typing import Any

import wit_world.imports.session as _session


def send(message: Any) -> None:
    """Send a message to the client.

    Strings are sent verbatim. Anything else is JSON-serialized via
    ``json.dumps`` (dicts, lists, numbers, bools, None). Objects with a
    ``model_dump_json`` method (e.g. future WASM-compatible
    pydantic-shaped classes) are handed off to it instead.

    ::

        session.send("plain text")
        session.send({"event": "tick", "n": 3})       # dict → JSON
        session.send([1, 2, 3])                       # list → JSON
    """
    if isinstance(message, str):
        _session.send(message)
        return
    if hasattr(message, "model_dump_json") and callable(message.model_dump_json):
        _session.send(message.model_dump_json())
        return
    _session.send(_json.dumps(message, default=str))


async def receive() -> str:
    """Receive a text message from the client."""
    result = await _session.receive()
    if result is None:
        raise RuntimeError("Session receive failed")
    return result


def send_file(data: bytes) -> None:
    """Send binary file data to the client."""
    _session.send_file(data)


async def receive_file() -> bytes:
    """Receive binary file data from the client."""
    result = await _session.receive_file()
    if result is None:
        raise RuntimeError("Session receive_file failed")
    return result

File: sdk/inferlet/javascript/src/session.ts
// Session functions — wraps pie:inferlet/session WIT interface.
// Handles communication with the remote user client.

import * as _session from 'pie:inferlet/session@0.3.0';

/** Sends a message to the remote user client.
 *
 * Strings are sent verbatim. Anything else is JSON-stringified
 * (objects, arrays, numbers, bools).
 *
 * ```ts
 * session.send("plain text");
 * session.send({ event: "tick", n: 3 });
 * session.send([1, 2, 3]);
 * ```
 */
export function send(message: unknown): void {
    if (typeof message === 'string') {
        _session.send(message);
    } else {
        _session.send(JSON.stringify(message));
    }
}

/** Receives a text message from the remote user client; `undefined` once
 *  the client has closed the connection.
 *
 *  Synchronous: a JS guest cannot lower the world's `async func` imports, so
 *  this goes through the host's blocking twin and the guest's task blocks
 *  until a message arrives. */
export function receive(): string | undefined {
    return _session.receiveBlocking();
}

/** Sends a file (binary data) to the remote user client. */
export function sendFile(data: Uint8Array): void {
    _session.sendFile(data);
}

/** Receives a file from the remote user client (synchronous; see `receive`). */
export function receiveFile(): Uint8Array | undefined {
    return _session.receiveFileBlocking();
}
